075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+This allows the user to not rely on the default Cruise Control goals configuration which would be common for any rebalance without overriding goals.
+They could define different goals based on the status of the Kafka cluster or skip hard goals in some scenarios for example.
+As already mentioned, specifying goals but more in general customizing the optimization proposal request is already possible in the current usage of a `KafkaRebalance` custom resource so it should be allowed for the auto-rebalancing as well.
+
+The idea of using a `KafkaRebalanceSettings` custom resource could be also extended to the manual rebalancing operations started by the user when they create a new `KafkaRebalance` custom resource.
+It would mean a change on the `KafkaRebalance` custom resource by adding a new `spec.settings` field to reference the rebalancing settings and deprecating all the other fields used today (i.e. `spec.goals`, ...).
+If this proposal is considered valid, doing such a change on the `KafkaRebalance` custom resource would help the proposal itself because the operator should not copy over the entire configuration from the settings resource to the rebalance resource but just set the settings reference.
+Of course, such a change would need a different proposal to be approved before this one.
+
+It is also worth mentioning that the additional `KafkaRebalanceSettings` custom resource doesn't bring any additional complexity to the operator in terms of watching, reacting and processing.
+The operator doesn't have any Kubernetes watch on the creation or update of such a resource.
+It can be only referenced by a `KafkaRebalance` custom resource so its creation doesn't trigger any action by the operator.
+The same happens when the `KafkaRebalanceSettings` custom resource is updated: no action is triggered for the operator, its updated content will be used only when a rebalancing is requested via a `KafkaRebalance` custom resource referencing those settings.
+
+### Extending the `Kafka.status` field with auto-rebalancing related state
+
+The auto-rebalancing, together with the scaling action, can take several reconciliations to finish.
+For this reason, taking into account cluster operator crashes as well, there is the need to save some state about it.
+The `Kafka` custom resource is extended by adding a new `status.cruiseControl.autoRebalance` field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # ...
+status:
+  cruiseControl:
+    autoRebalance:
+      state: # if a scaling operation is running or what's the state of the current rebalancing
+      add-brokers: # list with brokers' IDs to add (as part of the scaling up)
+      remove-brokers: # list with brokers' IDs to remove (as part of the scaling down)
```

Comment:
The `autoRebalance` section in `.spec` is extensible (defined as list). Why not do the same here?

Reply:
Not sure I got this comment. The `spec.autorebalance` has a list of "modes", so the `add-brokers` (or `remove-brokers`) mode value specify the mode for the rebalance custom resource. Here, `add-brokers` (or `remove-brokers`) is just a field for the list of brokers' IDs which are added (or removed).

Reply:
But it seems to use different format in different places. It seems like you should use the same in all of them.

Reply:
which places? can you be more precise with which fields are you comparing?
These `add-brokers` and `remove-brokers` fields are list of brokers' IDs.
They are like the `KafkaRebalance.spec.brokers` field, used to set a list of brokers' IDs, when you set `KafkaRebalance.spec.mode` to "add-brokers" or "remove-brokers". 
Are the same names which confuse you?

Reply:
```yaml
  cruiseControl:
    # ...
    autoRebalance:
      # using the same rebalancing settings when adding or removing brokers
      - mode: add-brokers
        settings: my-add-remove-brokers-rebalancing-settings
      - mode: remove-brokers
        settings: my-add-remove-brokers-rebalancing-settings
```

versus

```yaml
status:
  cruiseControl:
    autoRebalance:
      state: # if a scaling operation is running or what's the state of the current rebalancing
      add-brokers: # list with brokers' IDs to add (as part of the scaling up)
      remove-brokers: #
```

Reply:
> One has list and one is object with hardcoded modes.

But in the status they are not modes. Or better, they are the list of the brokers' IDs corresponding to that mode. They could be easily renamed `brokers-to-add` and `brokers-to-remove` (or `-being-added`, `-being-removed`). It's a way to track which are the brokers going to be added/remove during an autorebalancing process.

What we have in the spec is a list of modes with corresponding rebalancing settings. So when `add-brokers` use these settings, when `remove-brokers` use these others. There is no list of brokers in there.

Reply:
I do not understand how they would `not be modes`. They look like the status information stored when one of the auto-rebalance modes is triggers. So how come they are not tied to the modes?

Reply:
I mean ... when and how would you use it outside of the corresponding auto-rebalancing mode?

Reply:
I think I got what you meant now. If one of the spec is missing a mode, i.e. add-brokers, what's the meaning of having an hard-coded field in the status, i.e. add-brokers as an empty list of brokers' IDs.
So maybe you mean to have something like this in the status:

```yaml
status:
  cruiseControl:
    autoRebalance:
      state: # if a scaling operation is running or what's the state of the current rebalancing
      modes:
      - add-brokers: # list with brokers' IDs to add (as part of the scaling up)
      - remove-brokers: # list with brokers' IDs to remove (as part of the scaling down)
```

Reply:
Yes yes, I agree. It just took to me a lot of comments to understand what you meant ;-)

Reply:
Sorry for that, I probably badly explained it :-/